core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/CanonicalizeExpressionRewriter.java

```diff
@@ -294,20 +294,16 @@ public Expression rewriteFormat(Format node, Void context, ExpressionTreeRewrite
                     .addArgument(RowType.anonymous(argumentTypes.subList(1, arguments.size())), new Row(arguments.subList(1, arguments.size())))
                     .build();
         }
-    }
 
-    private static boolean isConstant(Expression expression)
-    {
-        // Current IR has no way to represent typed constants. It encodes simple ones as Cast(Literal)
-        // This is the simplest possible check that
-        //   1) doesn't require ExpressionInterpreter.optimize(), which is not cheap
-        //   2) doesn't try to duplicate all the logic in LiteralEncoder
-        //   3) covers a sufficient portion of the use cases that occur in practice
-        // TODO: this should eventually be removed when IR includes types
-        if (expression instanceof Cast && ((Cast) expression).getExpression() instanceof Literal) {
-            return true;
+        private boolean isConstant(Expression expression)
+        {
+            // Current IR has no way to represent typed constants. It encodes simple ones as Cast(Literal)
+            // This is the simplest possible check that
+            //   1) doesn't require ExpressionInterpreter.optimize(), which is not cheap
```

Comment:
With `isEffectivelyLiteral`, this is not quite true any more.

Reply:
How about this: https://github.com/trinodb/trino/pull/10663#discussion_r788778960 ?

Reply:
> How about this: [#10663 (comment)](https://github.com/trinodb/trino/pull/10663#discussion_r788778960) ?

Added tests in TestLogicalPlanner and TestInlineProjections.

> With isEffectivelyLiteral, this is not quite true any more.

it's still true. `ExpressionInterpreter.optimize()` is not cheap in general
while `ExpressionInterpreter.optimize()` is  used internally in `isEffectivelyLiteral`, the use is guarded with a check ensuring the call would be cheap.